### ISSUE TITLE
ERC 1155 like NFTs are not properly supported by metadata views

### ISSUE THREAD

**Issue by cybercent** [onflow/flow-nft#160]
## Instructions

Collections implementing  ERC1155 like NFTs are not properly supported (or too easy to break the standard).

### Problem

Using the same `Edition.name`, or a `nil` `Edition.name` leaves place for collisions.
A marketplace can't know if 2 NFTs are minted from the same `asset` as there is no `asset id` field.

TopShot does a good job by specifying an `asset Id` for each set in the `Edition.name` (example: `"Base Set: #1156"`),
but I think it's too easy for developers to shoot themselves in the foot by not specifying a name or specifying the same name for multiple editions, for example using only `"Base Set"` for multiple NFTs minted from different `assets`

```
pub struct Edition {

        /// The name of the edition
        /// For example, this could be Set, Play, Series,
        /// or any other way a project could classify its editions
        pub let name: String?

        /// The edition number of the object.
        /// For an "24 of 100 (#24/100)" item, the number is 24.
        pub let number: UInt64
```

**Comment by joshuahannan** [comment]
Can you elaborate a bit more on what an ERC 1155 like NFT is? It isn't clear from your description. Also, what would you propose to solve this?

**Comment by cybercent** [comment]
Yes, sorry.

ERC 1155 is a multi token standard on Ethereum [https://eips.ethereum.org/EIPS/eip-1155](https://eips.ethereum.org/EIPS/eip-1155) 

I was referring to tokens that can be created in multiple copies(editions) like  1/500.

NBATopShot is a good example of such a token.

The edition struct in MetadataViews needs the Name to be required OR a new required field (maybe edition_id ) to be able to refer to the same edition when no name is supplied.

**Comment by bluesign** [comment]
I think topshot is using wrong tbh, shouldn't it use `Editions` ?

AFAIK it should be something like: Series > Set > Play > Edition 

`Editions = [ Edition("Series: Season 1", 1, 3), Set("Set: Name Of Set", indexOfSet, maxSestInSeries1), Play("Play: Name of Play), indexOfPlay, maxPlayInThatSet) .... `

**Comment by joshuahannan** [comment]
yeah, Top Shot should use `Editions` like you've described.

I'm still a little confused about where the problem is. basing something on the edition name can be error prone, but basing it on an ID can also be error prone. If a project wants to make its editions distinct, it will make sure to make the names distinct, just like it would make its IDs distinct, right? And we're only comparing NFTs of the same type, so we don't have to worry about editions conflicting across types. 

The name not being required is for the case where there is only one edition. Are you seeing a lot of projects who have multiple editions but aren't giving them names?

**Comment by cybercent** [comment]
From a marketplace standpoint, we want to group moments created from the same asset. 
Instead of seeing the same moment multiple times, we display only one. 
This grouping was accomplished using the `asset_id` field, which is not visible in metadata views. 
The only way to do this right now is to use editions and group by edition name.

**Comment by cybercent** [comment]
Regarding the `attributes` I talked about here: https://github.com/onflow/flow-nft/issues/152 

Here's an example of bad practice in TopShot: the `SerialNumber` is included as a trait.

From a marketplace standpoint, we want to allow people to filter by trait and must generate a list of filters based on the traits. To get the exact idea of what I mean you can look at the filters (sidebar) present on any collection on OpenSea.

Because the serial numbers are unbounded, we must omit including them in the filters and manually clean the traits, as we did before metadata views were implemented.

Another bad practice on TS is using this type of trait name `playerFirstName`. 
This is the **exact trait name** users will see, why not use `Player first name`, or `Player First Name` ?

So in short, migrating from reading metadata manually and cleaning to reading metadata using metadata views will worsen a user's experience.

**Comment by joshuahannan** [comment]
It still sounds to me like this is a problem that needs to be addressed by individual projects, not by the NFT standard. Maybe I'm still not understanding fully what the issue is. I don't really see how we can enforce that traits have human readable names.

**Comment by joshuahannan** [comment]
I'm happy to share this with Top Shot to ask them to update their trait names and use better ones from now on though

**Comment by bluesign** [comment]
This is a great idea, as they are the first example most developers check when they start their flow journey.

**Comment by cybercent** [comment]
@joshuahannan 

Here is another example from Seussibles:

There are multiple NFTs with the same title "**Bowing Cat**", 
and same edition "**Standard edition**", 
but all with different images, so copies of a different asset.

There is no way to group NFTs that are copies of the same asset using the editions title field.

Arguably, this is a misuse of the `editions` info by the developer,
but Seussibles is an established project not a one-dev side project, 
so if this team didn't get how to use the metadata views the issue might be with the metadata views itself.

**Comment by joshuahannan** [comment]
I see what you mean, but I still don't see how we can enforce that kind of thing with the view standard. The whole point of metadata views is that we don't enforce how individual projects store and categorize their own views. I'm not sure what we can do here. I'm definitely open to suggestions though!

**Comment by bjartek** [comment]
> I'm happy to share this with Top Shot to ask them to update their trait names and use better ones from now on though

Changing this now will affect lots of other projects though FYI.

**Comment by cybercent** [comment]
For Seussibles the edition name is « Standard edition » but the NFTs are minted from 2 different assets.

It appears that some projects (TopShot, UFC etc) did manage to have unique editions by using an unique name while others just filled the field with what they assumed appropriate.

I don’t think there are many ways to think about this, a thing is either unique (1/1 like Flunks) or has copies(1/many like TopShot). When the thing has copies there should be a way to know that 2 NFTs are just copies of the same thing. Currently we can’t know that 2 NFTs are copies of the same thing. 

For Doodles there should have been a way to check on chain using metadata that 2 NFTs are copies of the same « blue socks » and are equal.

**Comment by bjartek** [comment]
For doodles they have a trait called `template_id` that says what template they are from. I was not aware that names for wearables where not unique across templates until after it was live. 

The idea for me for grouping has always been that if the name is the same then the edition numbers should diferentiate uniqueness. Now what edition to use determines how you want to group them. If there is only one edtion you can use that.

**Comment by joshuahannan** [comment]
I created this issue to work on a document that shows how views should be properly used? I'm considering closing this issue since it seems like the solution here is to make it clearer how to use the views properly instead of making a new view or changing and existing view. Does that sound ok?
